### DISCUSSION TITLE
Dependency Management

### DISCUSSION THREAD

**Discussion by sixolet** [monome/norns#1604] [General]
I'm interested in how we can manage dependencies for Norns scripts, in a way that isn't a heavy burden for script authors, and is mostly transparent for script users. 

## Uhh "user journeys" or something

Some use cases I am imagining, along with their current "solutions":

* The `toolkit` mod errors if the `matrix` mod is not installed, and is pointless if the `matrix` mod is not active. I would like to install `matrix` (if it's not installed) when installing `toolkit`.
  - I have documented this dependency, and users are expected to install both.
* Many scripts depend on particular UGens in sc, but UGens hate being double-installed. This should be easier.
  - @schollz has spent hella effort writing a script that checks all the likely places for the MI ugens and installs them if they're not already found, and then bundled this with several of his scripts.
  - Others have merely told you that you have to install the MI UGens.
* Many scripts have hard or soft dependencies on sc engines that are bundled with other scripts.
  - Many times the script author merely documents what other script has to be installed first.
  - Sometimes the dependency is soft, so the script author detects its absence and doesn't offer it as an option if it's not there. This can be a support burden.
* I have been using some of the same Lua libraries that I have written from a variety of scripts I write.
  - I have vendored them into each script, quick-and-dirty style. I guess I could use a git submodule, but that's clunky and annoying.
  - For very useful libraries like lattice, the dev team has accepted them in to core libs.
 * Anecdotally, it seems to me that engines or ugens being double-installed or not-installed is a significant portion of the overall Norns support burden on Lines, but I am an outsider to that.

## Thoughts toward solutions

* The Norns ecosystem already offers a "package repository" in the norns.community/Maiden database. This seems like a good start.
* We could define a simple JSON-based requirements file to declare dependencies by their maiden name in the top level dir of a script. Running an `;install` operation should automatically install the dependencies.
* We should mark things in the Maiden database that are intended to be library- or engine-only, so ppl don't expect them to be scripts.

## Challenges

* Once a script/maiden package is used as a library, an update to it could break anything that depends on it. This is the same problem we have in our current ad-hoc dependency management world, but encouraging dependencies would amplify it. 
  - Perhaps we could add a `library` or `stable api` tag to Maiden projects to indicate ones where the author commits to keeping the `lib` APIs backwards compatible. This requires library authors to be very good at backwards compatibility, though. And people will still depend on unstable stuff.
  - Classically we'd introduce a versioning system. Semver is pretty standard. The downside of this is that as soon as you start allowing depending on ranges of versions, you start having situations where you have insoluble version conflicts, if two scripts depended on different  and non-overlapping version ranges of the same library, and the different versions of the same library could not coexist (for example, they contained sc classes of the same name).
* The Maiden repo is already susceptible to "supply chain attacks" where a malicious user replaces the code for a package with bad code after it's accepted — but introducing dependencies increases the surface area. I don't think this is a major problem, but I just wanted to mention it.

## Not this project

* Lua that runs on installation (though that would be cool)
* Lua `Rocks`. Unfortunately, I don't think they are all that suited to projects that are naturally often a mix of lua and sc.

**Comment by tehn** [comment]
(thank you initiating this new place!)

we've spoken about this in the past and there are certainly some tangled issues with trying to "solve" the dependency problem.

my inclination:
- a straightforward/documented technique for detecting the presence of a package/lib/etc
- a system-level technique/function for aborting script load (ugh, or mod load?) and issuing an error message (on screen and in maiden)

this still relies on the script author to manage deps (which i think is not unreasonable) but hopefully sets a proper pattern to follow. i'm not inclined to weave in full dep-management magic.

(side note: i'm confused why ugen management has to  be so confusing, but this is likely better discussed as a separate issue)

we've had many discussions regarding `Rocks` in the past and have somewhat definitively decided against it, as it introduces a lot of surface area for failure.

**Comment by schollz** [comment]
the biggest dependency management hurdle I've encountered is definitely the ugen management.

for scripts its not a big deal - scripts fail and people report and we can pretty easily say why it fails since lua/maiden gives some hint at what lua dependency failed.  more importantly, when a script doesn't have a dependency it won't start, but it won't affect anything else.

ugens are trickier. SuperCollider will *fail to start* if it detects two ugens of the same name. this is problematic because it essentially bricks the norns: instead of one script not working, it makes *all* scripts unable to work.

as @sixolet mentioned I tried to develop [a script](https://github.com/schollz/paracosms/blob/main/paracosms.lua#L85-L120) in *paracosms* (which is using a couple 3rd party ugens) to detect ugens and not to copy them if they already exist. *paracosms* does this by storing its ugens in a `ignore` directory (which is [a magic directory name that SuperCollider will ignore](https://doc.sccode.org/Guides/UsingExtensions.html) when searching for ones to compile). *paracosms* moves ugens out of the `ignore` directory when a ugen is detected to be missing.

(detecting an existing ugen is tricky because there are three main locations they can be in:

-  `/usr/local/share/SuperCollider/Extensions`
 - `/home/we/dust/code`
- `/home/we/.local/share/SuperCollider/Extensions`

all of these are perfectly valid locations for putting 3rd-party ugen files, although the last one I think is considered the most canonical now.)

A solution to this ugen thicket might be a universal script to manage all the 3rd-party ugens, something like the monome-community scripts (I started [iterating on this idea](https://github.com/schollz/supercollider-plugins/)). another solution might look like detecting a SuperCollider FAIL and in the case of a duplicated class report it or quarantine it to an "ignore" directory so it doesn't brick the norns. anyways, open to any ideas so there can be fewer items in [this thread](https://llllllll.co/t/norns-error-supercollider-fail/45884/19)

**Comment by sixolet** [comment]
I’m happy with dependency management being on script authors, with some library support. I would like to remove it as a consideration for script users. 

A method to test for the presence of a library is nice, but I think even better would be a script-side hook into the maiden `;install` functionality, made idempotent so it’s not erroring the second time. Maybe the user should have to confirm? I don’t know.

**Comment by ngwese** [comment]
Couple of random comments that are hopefully helpful.

There is API for detecting whether a given mod is available/loaded or not. https://github.com/monome/norns/blob/main/lua/core/mods.lua#L90-L92 but I don't remember what the state of documentation is around mods beyond the original example code. If scripts are relying on functionality in mods then it would be good to test for them as a best practice.

Adding dependency management of some sort can certainly ease initial installation but without a whole bunch of additional work on the part of the script authors I'm not sure it would really help with maintaining stability over time. The more interconnected things get the more there is the possibility for script A and script B want different versions of a given dependency. Dependencies could be versioned with something like semver yet then it is entirely on the authors to know how to apply semver correctly. In order to avoid conflicts and interdependencies I would personally advocate the vendored approach (I personally use git submodules but I recognize those aren't for everyone). Vendored dependencies allow me to freely change library like code  I've written without breaking older scripts that I want to retain but don't have time to update.

One "dependency" not mentioned above that comes up frequently is which version of the norns software itself a given script assumes. There are plenty of examples where a new script is released, it doesn't work for someone, and the resolution is to have them update their device to the latest software. I see benefit in providing a function which could be placed at the top of a script which checks that the norns software is >= a given version like 20220310, if not fail gracefully with an on screen error.

The `;install` operation fails fast when it finds a directory which it would overwrite because it isn't the only means to "install" scripts on the device. Someone could have created a directory with the same name for a completely different purpose, something might have been manually copied into place via SMB/SFTP, there are no versions for zip archives necessarily so there is no guarantee that what is on disk is the same as what is in the archive being downloaded/cloned.

**Comment by tehn** [comment]
i'm very much in favor of of something like `norns.version_required("220101")` which could be added near the top of a script (ie, not within `init` so we catch it early), which would interface with the menu script loader and issue a sane error message.

even if you haven't dug into the menu code before this should be a pretty straightforward PR if someone wants to get into it

**Reply by tyleretters** [reply]
`norns.version_required` is a slam dunk.

could be extended to `norns.mod_required` and `norns.engine_required` too. we could even go so far as to have a lookup table of install links to be printed out for easy `;install` copy & paste.

(i'd love to have a `norns.mod_required("hecatomb")` for cci)

so basically as a phase one, we could skip over any programmatic dependency management and offload that to the user.

**Reply by robbielyman** [reply]
probably the version required should be an integer, since it will increase over time, so you can just implement the check with an inequality. I'll give this more basic thing a shot ~soon~ (later today?) unless someone else wants to do it ^-^

**Reply by tehn** [reply]
we already have an "engine not found" mechanism

makes sense that the version should be an integer for comparison

**Reply by dndrks** [reply]
if it helps, [this](https://github.com/dndrks/cheat_codes_2/commit/88dca33f2527e2e54a9e507f86e36b49a186090f) has been working for cheat codes 2.
generally, i really also dig the idea of an on-screen UI message helping differentiate between 'hey, this requires a different norns version, so run SYSTEM > UPDATE' and 'hey, reflash norns'. it'd really help keep pre-`220306` folks who aren't connecting to maiden while updating from ending up in the ['i ran UPDATE but it didn't update, what happened?'](https://llllllll.co/t/cheat-codes-2-rev-221007-lts9-more-important-fixes/38414/1521) cycle.

**Reply by robbielyman** [reply]
I gave it a shot in #1613. It occurs to me that while @dndrks's idea is super rad, it's not really possible to implement on a system level as far as I can tell, since in order for norns to know that a new image was necessary, an update newer than the image needs to be on norns. Nevertheless, I included a suggestion about checking for a new disk image in the maiden output.

**Comment by tehn** [comment]
thank you!

yes we frequently run into this issue--- we want an error alert for something but we can't retroactively add it to previous versions without having an update. i hit this wall dozens of times trying to refine the actual update process :)

**Comment by robbielyman** [comment]
Regarding UGens, here's a pattern that I shared on Discord that will allow a script to install a UGen shipped with the script in an `/ignore` directory (thanks @schollz for rubber-ducking my "why isn't this working"!) in what I'd like to consider the canonical location for it, `/home/we/.local/share/SuperCollider/Extensions/UGenName/`. The drawback is of course that all other scripts installing the UGen must do so in the same location, but this does allow the script to load, install the UGens, and show the error message on the Norns screen and hopefully avoid the `duplicate engines` error.
```lua
-- script
local extensions = "/home/we/.local/share/SuperCollider/Extensions"
engine.name = util.file_exists(extensions .. "/UGenName/UGenName.sc") and "EngineName" or nil

UI = require("ui")

function init()
  Needs_Restart = false
  local ugen_files = {"UGenName.sc", "UGenName_scsynth.so"}
  for _,file in pairs(ugen_files) do
    if not util.file_exists(extensions .. "/UGenName/" .. file) then
      util.os_capture("mkdir " .. extensions .. "UGenName")
      util.os_capture("cp " .. norns.state.path .. "/ignore/" .. file .. " " .. extensions .. "/UGenName/" .. file)
      print("installed " .. file)
      Needs_Restart = true
    end
  end
  Restart_Message = UI.Message.new{"please restart norns"}
  if Needs_Restart then redraw() return end
  -- rest of init()
end

function redraw()
  if Needs_Restart then
    screen.clear()
    Restart_Message:redraw()
    screen.update()
    return
  end
  -- rest of redraw()
end
```

**Reply by schollz** [reply]
Thanks @ryleelyman, I'm using this right now for a script and it worked great.

while on the subject of compiled UGens - there was another "gotcha" I found when paracosms started being downloaded by people with fates. I was distributing compiled UGens and UGens I distributed were compiled on my norns shield. However it came to be a problem when people installed them on their "fates" because fates is currently behind in its GLIBC (2.28 since its Debian 10) so they would get errors. To fix the problem I asked okyeron to recompile everything on their fates, and then I distributed those instead (since GLIBC is backwards compatible this works with the newer GLIBC on norns). Its not a very sustainable solution...since I don't have a fates or a 32-bit Debian 10 system I can't really re-compile anything if anything changes.

Another possibility to circumvent this particular fates vs norns problem is to compile the UGen as part of the process of installing (which is what I'm doing with fverb in zxcvbn now). The above script enabled that for me (each user compiles the fverb code on their own norns, whatever version it is).

**Reply by robbielyman** [reply]
hmmmm. that's quite a gotcha. the UGens I made need `cmake`, at least as written, which I needed to build from source to compile on my norns...

does this mean that the scripts that use my UGens will fail on fates? if so, I'm kind of surprised nobody with one has tried bitters, lamination or schicksalslied 😅

**Reply by schollz** [reply]
no not necessarily! for all I know it only happened with a particular UGen from f0plugins on fates ([i.e. here](https://llllllll.co/t/paracosms/57350/15)). but I can't say I know much about this - I don't understand either fates or SuperCollider compilation entirely. I know that this didn't happen with other plugins (like `PortedPlugins`). your UGens may be okay! but its still a gotcha for some...

**Comment by tehn** [comment]
i'd be up for considering a core component (ie, included in the main norns codebase) which is an install helper.

this way a script could achieve the above simply by requiring an engine lib ie require Polysub, which would in turn do a norns.require.ugen (or something) which could check for the ugen and if needed install it from the local folder and issue a restart request, or if not found sound an alarm

we may need to have a correlate on-startup check to debug/fix/alert the user of misplaced ugens--- though this is a broader issue when it comes to .sc files (class collision, which is harder to solve)

**Reply by robbielyman** [reply]
I'll give this first piece a shot. maybe an `engine.ugens = {'UGenName1', 'UGenName2'}` model in the Lua side someplace outside of `init` would work?

**Reply by robbielyman** [reply]
gave it a try in #1614. Managed to create a persistent `Supercollider fail` on one of my norns but not the other, which leads me to suspect my PR is not the cause, but it's worth looking out for.
